js/components/dashboard/activity-answers.js

```diff
@@ -51,3 +53,30 @@ export default class ActivityAnswers extends PureComponent {
     );
   }
 }
+
+function mapStateToProps(state, ownProps) {
+  const questions = ownProps.activity.get("questions") || [];
+  const questionIds = questions.map(question => question.get("id"));
+  const answerTreeList = getAnswerTrees(state).toList();
+
+  return {
+    // This computes an answers property here by finding all of the answers
+    // for the questions we were passed.
+    // this is only used to compute a score, so it might be better to just compute
+    // the score here instead of passing around answers
+    answers: answerTreeList.filter((answer) => {
+      // This is not very efficient because we are iterating over every questionId in the
+      // the activity for each answer in the activity
+      return questionIds.includes(answer.get("questionId")) &&
```

Comment:
Would an initial pass over the `answerTreeList` to build a map from `questionId` to answers make sense? Or perhaps a map from `userId:questionId` to answers, which would be a larger map but involve less looping.

Reply:
Yes some kind of map here is what we need. I think ideally `getAnswerTrees(state)` would return a map or a map of maps so there is no need for filtering and iterating, just a simple lookup.  But I'm not sure if that is good practice. All of the selectors in https://github.com/concord-consortium/portal-report/blob/master/js/selectors/report-tree.js seem to return arrays.  In the example scenario above, that ought to cut down the 207,360,000 questionId checks to loop over 14,400 which creates maps.